### PR TITLE
feat(query, core): Introduce query session to store mutable state per-query

### PR DESCRIPTION
**Pull request by vishramachandran**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

QuerySession object can be used to store per-query lock state, trace information etc.

### CODE REVIEW

File: core/src/main/scala/filodb.core/store/package.scala [hunk 1]
```diff
@@ -145,8 +146,9 @@ package object store {
     def scanRows(dataset: Dataset,
                  columnIDs: Seq[ColumnId],
                  partMethod: PartitionScanMethod,
-                 chunkMethod: ChunkScanMethod = AllChunkScan): Iterator[RowReader] =
-      source.scanPartitions(dataset.ref, columnIDs, partMethod, chunkMethod)
+                 chunkMethod: ChunkScanMethod = AllChunkScan,
+                 querySession: QuerySession = QuerySession.forTestingOnly): Iterator[RowReader] =
```
Comment (broneill) [inline core/src/main/scala/filodb.core/store/package.scala:150]: Should the scanRows method be tagged as "for testing only" too?

File: core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala [hunk 2]
```diff
@@ -410,6 +410,8 @@ class TimeSeriesShard(val ref: DatasetRef,
     */
   val tracedPartFilters = storeConfig.traceFilters
 
+  protected val shardReclaimLock = new StampedLock
```
[the remark this answers is not in the document]
Comment (vishramachandran) [reply core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala]: yes - that was not intended . removed.

File: core/src/main/scala/filodb.core/query/QueryContext.scala [hunk 3]
```diff
@@ -57,3 +57,16 @@ object QueryContext {
     simpleMapSpreadFunc(shardKeyNames.asScala, spreadAssignment, defaultSpread)
   }
 }
+
+case class QuerySession(qContext: QueryContext,
```
[the remark this answers is not in the document]
Comment (vishramachandran) [reply core/src/main/scala/filodb.core/query/QueryContext.scala:61]: I had it that way initially - then it occurred to me that much more state would be added - like traces, query stats etc. If you have strong reasons to not have it, I'll just leave it as is.

File: core/src/main/scala/filodb.core/store/package.scala (hunk 1 above)
Comment (vishramachandran) [reply core/src/main/scala/filodb.core/store/package.scala:150]: Yes - it actually should. Dont know the original intent. Will just leave it.

File: core/src/main/scala/filodb.core/query/QueryContext.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (broneill) [reply core/src/main/scala/filodb.core/query/QueryContext.scala:61]: It's just a one-to-one relationship? The level of indirection is just that -- I don't see any value in creating another object and pointer hop. Is the goal to ensure that session is case class? Is it that critical? Are we applying pattern matching to it?

File: core/src/main/scala/filodb.core/store/package.scala (hunk 1 above)
Comment (broneill) [reply core/src/main/scala/filodb.core/store/package.scala:150]: I think a comment will help. I see "forTestingOnly" in the code and this isn't mirrored in the scanPartitions doc or method name.

**Pull request merged**